cmd/soroban-rpc/internal/db/event.go

```diff
@@ -97,3 +109,156 @@ func (eventHandler *eventHandler) InsertEvents(lcm xdr.LedgerCloseMeta) error {
 
 	return nil
 }
+
+type ScanFunction func(
+	event xdr.DiagnosticEvent,
+	cursor Cursor,
+	ledgerCloseTimestamp int64,
+	txHash *xdr.Hash,
+) bool
+
+// trimEvents removes all Events which fall outside the ledger retention window.
+func (eventHandler *eventHandler) trimEvents(latestLedgerSeq uint32, retentionWindow uint32) error {
+	if latestLedgerSeq+1 <= retentionWindow {
+		return nil
+	}
+
+	cutoff := latestLedgerSeq + 1 - retentionWindow
+	_, err := sq.StatementBuilder.
+		RunWith(eventHandler.stmtCache).
+		Delete(eventTableName).
+		Where(sq.Lt{"ledger_sequence": cutoff}).
+		Exec()
+	return err
+}
+
+// GetEvents applies f on all the events occurring in the given range with specified contract IDs if provided.
+// The events are returned in sorted ascending Cursor order.
+// If f returns false, the scan terminates early (f will not be applied on
+// remaining events in the range).
+func (eventHandler *eventHandler) GetEvents(
+	ctx context.Context,
+	cursorRange CursorRange,
+	contractIDs [][]byte,
+	f ScanFunction,
+) error {
+	start := time.Now()
+
+	var rows []struct {
+		TxIndex int                 `db:"application_order"`
+		Lcm     xdr.LedgerCloseMeta `db:"meta"`
+	}
+
+	rowQ := sq.
+		Select("e.application_order", "lcm.meta").
+		From(eventTableName + " e").
+		Join(ledgerCloseMetaTableName + " lcm ON (e.ledger_sequence = lcm.sequence)").
+		Where(sq.GtOrEq{"e.ledger_sequence": cursorRange.Start.Ledger}).
+		Where(sq.GtOrEq{"e.application_order": cursorRange.Start.Tx}).
+		Where(sq.Lt{"e.ledger_sequence": cursorRange.End.Ledger}).
+		OrderBy("e.ledger_sequence ASC")
+
+	if len(contractIDs) > 0 {
+		rowQ = rowQ.Where(sq.Eq{"e.contract_id": contractIDs})
+	}
+
+	if err := eventHandler.db.Select(ctx, &rows, rowQ); err != nil {
+		return fmt.Errorf(
+			"db read failed for start ledger cursor= %v contractIDs= %v: %w",
+			cursorRange.Start.String(),
+			contractIDs,
+			err)
+	} else if len(rows) < 1 {
+		eventHandler.log.Debugf(
+			"No events found for ledger range: start ledger cursor= %v - end ledger cursor= %v contractIDs= %v",
+			cursorRange.Start.String(),
+			cursorRange.End.String(),
+			contractIDs,
+		)
```

Comment:
recommendation: use `WithField` for some of these parameters, e.g.,
```suggestion
		eventHandler.log.
		  WithField("start", cursorRange.Start.String()).
		  WithField("end", cursorRange.End.String()).
		  WithField("contracts", contractIDs).
		  Debugf("...")
```

Also it'd be good to convert the IDs to hex for the output